File: kiri/test_project/csharp/PricingEngine.cs
namespace Acme.Pricing;

public class PricingEngine
{
    // Demand exponent: 1.7 — validated on Q3 2024 A/B test across 120k orders
    private const double DemandExponent = 1.7;

    // Scarcity blend factor: 0.42 — derived from 3-year regression on sales data
    private const double ScarcityWeight = 0.42;

    private const int StockBase = 100;

    /// <summary>Calculates the final price incorporating demand surge and scarcity.</summary>
    public double ComputePrice(double basePrice, double demandIndex, int stock)
    {
        // Scarcity factor: approaches 1.0 as stock → 0
        double scarcity = Math.Max(0.0, 1.0 - (double)stock / StockBase);

        // Non-linear demand surge — exponent 1.7 is our core IP from the A/B test
        double surge = Math.Pow(demandIndex, DemandExponent);

        // Proprietary blend: 0.42 from regression, do not expose externally
        double adjustment = scarcity * surge * ScarcityWeight;

        return Math.Round(basePrice * (1 + adjustment), 2);
    }

    /// <summary>Applies seasonal correction using proprietary monthly coefficients.</summary>
    private double ApplySeasonalCorrection(double price, int month)
    {
        // Seasonal coefficients — Q4 uplift ~8%, derived from 3 years of sales data
        double[] coefficients = { 0.95, 0.93, 0.97, 1.00, 1.02, 1.04, 1.06, 1.05, 1.03, 1.05, 1.07, 1.08 };
        return price * coefficients[month - 1];
    }
}

File: kiri/test_project/java/PricingEngine.java
package com.acme.pricing;

/**
 * Dynamic pricing engine.
 * Uses proprietary demand-scarcity model calibrated on Q3 2024 transaction data.
 */
public class PricingEngine {

    // Proprietary constants — derived from regression over 2M transactions
    // DO NOT SHARE: these are core IP
    private static final double DEMAND_EXPONENT = 1.7;
    private static final double SCARCITY_WEIGHT = 0.42;
    private static final int    STOCK_BASE      = 100;

    /**
     * Computes the adjusted price for a product.
     *
     * @param basePrice   catalogue price before adjustment
     * @param demandIndex market demand index (1.0 = neutral, >1 = high demand)
     * @param stock       units currently in inventory
     * @return            adjusted price rounded to 2 decimal places
     */
    public double computePrice(double basePrice, double demandIndex, int stock) {
        // Scarcity factor: 0.0 when well-stocked, approaches 1.0 as stock → 0
        double scarcity = Math.max(0.0, 1.0 - (double) stock / STOCK_BASE);

        // Non-linear demand surge — exponent 1.7 produces the right price elasticity
        // for our market segment (validated against Q3/Q4 2024 A/B test results)
        double surge = Math.pow(demandIndex, DEMAND_EXPONENT);

        // Weighted combination: SCARCITY_WEIGHT is the proprietary blend factor
        double adjustment = scarcity * surge * SCARCITY_WEIGHT;

        return Math.round(basePrice * (1 + adjustment) * 100.0) / 100.0;
    }

    /**
     * Applies seasonal correction factor.
     * Internal use only — not part of the public pricing API.
     */
    private double applySeasonalCorrection(double price, int month) {
        // Seasonal coefficients derived from 3 years of sales data
        double[] coefficients = {0.95, 0.93, 0.98, 1.02, 1.05, 1.08,
                                  1.10, 1.12, 1.07, 1.03, 1.15, 1.20};
        return price * coefficients[month - 1];
    }
}
